### PR TITLE
Named cache

### LINKED ISSUE #106: Support for NamedCache
Hello there, I’m looking to use `play-redis` in an upcoming project, but have some modules that depend on Play’s `NamedCache` [feature](https://www.playframework.com/documentation/2.6.x/ScalaCache#Accessing-different-caches). 

Is support for `NamedCache` on your road map? Would you be open to a pull request implementing this support?

Thanks for the excellent module!

### PR DESCRIPTION
**Pull request by KarelCemus**
This PR redesigns configuration to fully implement named caches. Fixes #106.

This is intended for version 2.x.x without backward compatibility of the configuration. There will be warnings and migration guide to allow more advanced usage.

@pcejrowski @rval May I ask you to review this PR, especially the `reference.conf` file? It should be final and stable. Modules and Components will change for sure,  some other binding-related classes might change too but the configuration proposal is final.

### CODE REVIEW

**Comment by pcejrowski** [comment]
Hey, sure. I'm going to review reference.conf tomorrow and I'll take a deeper look on the other files  during the weekend. I'm short with time earlier.

**Comment by KarelCemus** [comment]
Great, thanks! No rush. And please ignore tests, they are about to be fully rewritten.

File: src/main/resources/reference.conf [hunk 1]
```diff
@@ -1,41 +1,127 @@
...
-  ]
-  # connection timeout in milliseconds
-  timeout:    1s
+play.cache.redis: {
+
+  # list of named caches available in the application
+  #
+  # for more details see https://www.playframework.com/documentation/2.6.x/ScalaCache#Accessing-different-caches
+  #
+  instance {
```
Comment (checat) [inline src/main/resources/reference.conf]: I would make it `instances`

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -1,41 +1,127 @@
...
+
+    # definition of the default cache,
+    # its name is defined in 'play.cache.defaultCache'
```
Comment (checat) [inline src/main/resources/reference.conf]: `play.cache.redis.default-cache` is defined below, but `play.cache.defaultCache` is mentioned here

**Comment by KarelCemus** [comment]
Thank you @checat for your feedback, it's fixed.

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -1,41 +1,127 @@
...
+  #  - 'cluster' mode indicates use of 'cluster' variable defining nodes
+  #  - 'connection-string' mode is usually used with PaaS as setup by the
+  #      environment. It consideres 'connection-string' property.
+  #  - 'custom' indicates that the user supplies his own RedisInstance configuration
+  #
+  # Default value is 'standalone'
+  #
+  # note: this is global definition, can be locally overriden for each
+  # cache instance. To do so, redefine this property
+  # under 'play.cache.redis.instance.instance-name.this-property'.
```
Comment (rval) [inline src/main/resources/reference.conf]: Here too.

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -1,41 +1,127 @@
...
+  #
+  source:           standalone
+
+  # synchronization timeout applies only with synchronous API. It does not
+  # apply with asynchronous API and you are free to ignore it. This value
+  # is used with `Await.result`.
+  #
+  # note: this is global definition, can be locally overriden for each
+  # cache instance. To do so, redefine this property
+  # under 'play.cache.redis.instance.instance-name.this-property'.
```
Comment (rval) [inline src/main/resources/reference.conf]: And here.

**Comment by KarelCemus** [comment]
@rval Thanks for your feedback, I'm sorry for my sloppiness. It should be fixed.

I'll wait couple more days for the last feedback and then merge the PR.

File: src/main/resources/reference.conf [hunk 5]
```diff
@@ -1,41 +1,127 @@
...
+      ##########################
+      # Standalone mode
+      ##########################
+
+      # redis server: host
+      host:       localhost
+      # redis server: port
+      port:       6379
+      # redis server: database number
+      database:   1
```
Comment (pcejrowski) [inline src/main/resources/reference.conf]: I realize that it may break some people's usage.

**Comment by KarelCemus** [comment]
@pcejrowski Thanks for your feedback. I am worried about the complexity of the configuration. I hoped for simpler settings but it conflicts with the need for expressing complex setups.

I mean that small apps with just single redis cache will have quite complex settings I think (comparing to version 1.x). 

Do you have any ideas to improve the situation or do you think it is alright to keep it this way?

File: src/main/resources/reference.conf (hunk 5 above)
Comment (KarelCemus) [inline src/main/resources/reference.conf]: Honestly, I don't mind that much for backward compatibility right now as there is none right now. The change is too big to keep it compatible I think.

But on the other hand, the complexity of the new configuration is worrying me, as I posted couple secs ago.

Comment (pcejrowski) [inline src/main/resources/reference.conf]: Sure. That's the right point of time to make any breaks.

**Comment by pcejrowski** [comment]
True.
(Thinking aloud:)
You can assume, that the usage is with single server XOR cluster, thus mode parameter and different configuration subtree might be valuable. However, it may create higher configuration reading complexity.

**Comment by KarelCemus** [comment]
That's true, basically, I wanted to do it at the beginning. But then there raised two questions:

1. Would it make it more simple or more complex to comprehend? There would have to be two documentations, two examples, etc.
2. What to do when the user declares both types of the configuration. Furthermore, one of them should be active by default to ease the initial use thus the other one would always conflict with the first one.

I concluded that it would make the things much more difficult to understand instead of easing the use. 

What do you think?

**Comment by pcejrowski** [comment]
Just realized `source` parameter

**Comment by KarelCemus** [comment]
> Just realized source parameter

Yes, right. I don't like it but didn't figure out any other way to overcome the limitations I mentioned and support multiple types of setup.

But still, it is very complex for the simple situations. Consider this:

```
play.cache.redis.instances.play {
  host: ...
  port: ...
  database: ...
}
```

I think this is the most simple use case but I very dislike the key, it does not seem intuitive at first.

**Comment by pcejrowski** [comment]
Another possible solution I can think of is as follows:

for single server:
```
play.cache.redis {
    host: ...
    port: ...
}
```

for cluster
```
play.cache.redis {
    named-cache: true
    instances.play {
   ...
   }
}
```

with providing named-cache: false, which simplifies 99% of use-cases, right? (Named cache was requested after a long time of successful usage of `play-redid` in many projects.)

**Comment by KarelCemus** [comment]
So basically you propose another "`source`" parameter switching the configuration trees. 

By default, when `name-cache` is false, `play.cache.redis.instances.play` is projected to `play.cache.redis`. Otherwise it is ignore.

I think that might work and would not complicate either the comprehension or the implementation as it is basically just projection. Furthermore, as I proposed the settings a bit "recursively", it works smoothly with that. In the case of `name-cache` false the default settings would equal to settings of the default cache.

Furthermore, in 99% cases, the migration would be quite simple I believe.

I really like it, thanks!

**Comment by pcejrowski** [comment]
Exactly. 
The bad thing here is defined by keyword: **another**.. I don't like many flags/toggles solutions, but it might be reasonable here. Have you thought about any other potential dimensions (server/cluster/connection string, named/not-named, etc.) that may come out in the future? 

I tried to think of it right now and it seems, that with those two flags you cover all the specificities.

**Comment by KarelCemus** [comment]
> I tried to think of it right now and it seems, that with those two flags you cover all the specificities.

Well, personally, I use it in a very simple scenario, so I'm not aware of all possibilities of the redis. But I cannot think of anything I know what would not be supported by these two switches.

It supports standalone/cluster/cloud as well as single cache/multiple named caches.

I can image introduction of other values of `source` property, but the mechanism and the configuration it self would remain intact, so I should cover everything available right now.

**Comment by pcejrowski** [comment]
That sounds pretty good. One main knob per dimension, that turns on subtree. That would be really straightforward to use it that way. Familiar, thus intuitive.

Maybe @checat @rval @solicode can add more input.

**Comment by rval** [comment]
Apologies for the late response. 

I think both options work well, but I have a slight preference for the original `instances.name` approach over the alternative; I think having `named-cache` change the configuration tree is more confusing than adding `instances.play` into its hierarchy. I also like the consistency in configuring either `instances.play` or `instances.some-name`.

I assume this new release will break backwards compatibility one way or another given the other changes planned. I think most integrators would accept some minor work to get on the latest release with its improvements, particularly if those are documented in `UPGRADE-2.0.md`.

Another thought—if `instances.default` is the actual default, we could remove the `default-cache` parameter (potential downside, what if someone actually wants a named cache of "default"?). I also think `play.cache.redis.instances.default` is quite clear for new and existing users.

```yaml
play.cache.redis {
    instances {
        default: { # No need for default-cache since this is the default.
            host: "some.host",
            port: 1234
        },
        session-cache: {
            host: "some.other.host",
            port: 4321
        }
    }
}
```

Like I said, just personal preference. Either approach works for me.

**Comment by johnsiu** [comment]
If two named caches were configured to point to the same Redis host, port, and database, then it would be possible to have key collision.

Is it possible to add an option for named cache to implicitly prepend the name to the key?

akin to play2-memcached
https://github.com/mumoshu/play2-memcached/blob/master/plugin/src/main/play_2.6/com/github/mumoshu/play2/memcached/MemcachedCacheApi.scala#L35

**Comment by rval** [comment]
@johnsiu that crossed my mind also, but I wonder if a warning in the comments of `reference.conf` and/or `README.md` might be sufficient? [This post](http://www.mikeperham.com/2015/09/24/storing-data-with-redis/) has lists a few reasons to avoid namespacing in Redis.

**Comment by johnsiu** [comment]
> @johnsiu that crossed my mind also, but I wonder if a warning in the comments of reference.conf and/or README.md might be sufficient? This post lists a few reasons to avoid namespacing in Redis.

@rval
Here are the two reasons listed by the post:

> 1. Namespacing increases the size of every key by the size of the prefix.

This seems quite minor to me, assuming namespace is small.

> 2. You don't get to tune Redis for the individual needs of "cache" and "transactional"

I am suggesting to add namespace as an option, which can be enabled/disabled for each named cache. We are still able to configure and tune individual cache as needed. 

Namespace works quite naturally with named cache since the "name" of the cache itself is the namespace. 
The common use case is that a single Play application has a single cache need, but has logical division into modules, which are developed by different teams.

The different options to avoid key collision are:
1. spin up multiple Redis instances.
Hard to find a business justification for the operational overhead/cost.
2. use different Redis databases.
Deprecated and not supported by Redis Cluster.
3. co-ordination between teams.
low tech and error-prone.
4. use namespace.

I would argue that 4 is quite a reasonable option.

**Comment by KarelCemus** [comment]
Hi guys, thanks for your contribution.

@johnsiu I created a new ticket #118 for this discussion to keep this thread focused.

> having named-cache change the configuration tree is more confusing than adding instances.play into its hierarchy

Well it depends, I think not everybody is familiar with named caches so force them to "know them" does not seem right, though the configuration is still simple. I think it might be intimidating for new users and could increase the initial barrier. Thus lowering it by keeping the default config simple in 95% cases seems right.

> I assume this new release will break backwards compatibility one way or another given the other changes planned. 

That's true, but I think it will break only configuration compatibility. Otherwise, it should be source compatible for regular users (not those who somehow extend or modify the core).

> I think most integrators would accept some minor work to get on the latest release with its improvements, particularly if those are documented in UPGRADE-2.0.md.

I agree.

> Another thought—if instances.default is the actual default, we could remove the default-cache parameter (potential downside, what if someone actually wants a named cache of "default"?). I also think play.cache.redis.instances.default is quite clear for new and existing users.

Well, it occurred to me too, but I decided to remain consistent with Play, which uses `play` cache as a default in its `EhCache` implementation.

**Comment by rval** [comment]
@KarelCemus all your counterpoints are totally reasonable. Like I said, just my own personal preference, certainly not a deal breaker. 

Please let me know if there's anything I can do to help.

**Comment by KarelCemus** [comment]
@rval Thanks for your feedback, really appreciate it. Unfortunately, I must apologize for this pending PR, I have really busy 3 weeks and had no time to deal with this. One week left and it should be better again. Then I'm about to complete this PR. I have a pretty clear vision how to finish it.

Thanks for your patience.

**Pull request merged**